DB/src/main/java/io/deephaven/db/v2/sources/chunkcolumnsource/CharChunkColumnSource.java

```diff
@@ -0,0 +1,165 @@
+                if (offset == 0 && length == charChunk.size()) {
+                    return charChunk;
+                }
+                return ((ChunkGetContext) context).resettableCharChunk.resetFromChunk(charChunk, offset, length);
+            }
+        }
+        return getChunkByFilling(context, orderedKeys);
+    }
+
+    @Override
+    public void fillChunk(@NotNull final FillContext context, @NotNull final WritableChunk<? super Attributes.Values> destination, @NotNull final OrderedKeys orderedKeys) {
+        final MutableInt searchStartChunkIndex = new MutableInt(0);
+        destination.setSize(0);
+        orderedKeys.forAllLongRanges((s, e) -> {
+            while (s <= e) {
+                final int chunkIndex = getChunkIndex(s, searchStartChunkIndex.intValue());
+                final int offsetWithinChunk = (int) (s - firstOffsetForData.get(chunkIndex));
+                Assert.geqZero(offsetWithinChunk, "offsetWithinChunk");
+                final CharChunk<? extends Attributes.Values> charChunk = data.get(chunkIndex);
+                final int chunkSize = charChunk.size();
+                final long rangeLength = e - s + 1;
+                final int chunkRemaining = chunkSize - offsetWithinChunk;
+                final int length = rangeLength > chunkRemaining ? chunkRemaining : (int) rangeLength;
+                Assert.gtZero(length, "length");
+                final int currentDestinationSize = destination.size();
+                destination.copyFromChunk(charChunk, offsetWithinChunk, currentDestinationSize, length);
+                destination.setSize(currentDestinationSize + length);
+                searchStartChunkIndex.setValue(chunkIndex + 1);
+                s += length;
+            }
+        });
+    }
+
+    @Override
+    public void fillPrevChunk(@NotNull final FillContext context, @NotNull final WritableChunk<? super Attributes.Values> destination, @NotNull final OrderedKeys orderedKeys) {
+        // immutable, so we can delegate to fill
+        fillChunk(context, destination, orderedKeys);
+    }
+
+    /**
+     * Given an index within this column's address space; return the chunk that contains the index.
+     *
+     * @param start the data index to find the corresponding chunk for
+     * @return the chunk index within data and offsets
+     */
+    private int getChunkIndex(final long start) {
+        return getChunkIndex(start, 0);
+    }
+
+    /**
+     * Given an index within this column's address space; return the chunk that contains the index.
+     *
+     * @param start      the data index to find the corresponding chunk for
+     * @param startChunk the first chunk that may possibly contain start
+     * @return the chunk index within data and offsets
+     */
+
+    private int getChunkIndex(final long start, final int startChunk) {
+        int index = firstOffsetForData.binarySearch(start, startChunk, firstOffsetForData.size());
```

Comment:
Should we insist that all chunks except the last one have the same size, so we don't have to do a binary search? I guess it would make this tool less general.

Reply:
No, but I added an escape for when you march through in order.